source/extensions/filters/network/postgres_proxy/postgres_filter.cc

```diff
@@ -186,12 +187,53 @@ void PostgresFilter::processQuery(const std::string& sql) {
   }
 }
 
-void PostgresFilter::doDecode(Buffer::Instance& data, bool frontend) {
+bool PostgresFilter::onSSLRequest() {
+  if (!config_->terminate_ssl_) {
+    // Signal to the decoder to continue.
+    return true;
+  }
+  // Send single bytes 'S' to indicate switch to TLS.
+  // Refer to official documentation for protocol details:
+  // https://www.postgresql.org/docs/current/protocol-flow.html
+  Buffer::OwnedImpl buf;
+  buf.add("S");
+  // Add callback to be notified when the reply message has been
+  // transmitted.
+  read_callbacks_->connection().addBytesSentCallback([=](uint64_t bytes) -> bool {
+    // Wait until 'S' has been transmitted.
+    if (bytes >= 1) {
```

Comment:
Sorry, this is mostly my ignorance on the protocol, here: https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.11 we expect 'S' and 'N'? Where in the code that we handle 'N' response?

Reply:
Because here we need to mimic Postgres server sending back to client that we support SSL so the client is able to start the handshake.

And if not exists a transport socket then the connection will be finished.

Reply:
@dio Thanks for reviewing. The logic here is that the postgres filter "captures" the initial request from the client and does not allow it to reach the server. Then it sends a response 'S', which means "let's switch to SSL". After the response has been transmitted, it calls starttls socket to switch to TLS/SSL. The call to switch to TLS may fail (configuration, wrong certs, etc). Theoretically when switch to TLS fails, the postgres filter should transmit 'N', which means "no, I do not allow to switch to TLS". But the filter has already transmitted 'S' and if switch to TLS fails, postgres filter will close the TCP connection to the client.

In other words, we do not send 'N', but close connection when things go wrong.

Reply:
@dio this [image](https://user-images.githubusercontent.com/1239643/80261846-e5129d80-868b-11ea-995d-0a55ae96ce87.png) from original issue describes this behavior. For sure it don't cover the STARTTLS stuff because at that time we have no idea about all Envoy internals we need to change/improve to implement this feature.

Reply:
Thank you for the explanation!